pkg/tfvars/openstack/rhcos_image.go

```diff
@@ -0,0 +1,140 @@
+	}
+	defer f.Close()
+
+	opts := clientconfig.ClientOpts{
+		Cloud: cloud,
+	}
+
+	conn, err := clientconfig.NewServiceClient("image", &opts)
+	if err != nil {
+		return err
+	}
+
+	imageCreateOpts := images.CreateOpts{
+		Name:            imageName,
+		ContainerFormat: "bare",
+		DiskFormat:      "qcow2",
+		Tags:            []string{fmt.Sprintf("openshiftClusterID=%s", clusterID)},
+		// TODO(mfedosin): add Description when gophercloud supports it.
+	}
+
+	img, err := images.Create(conn, imageCreateOpts).Extract()
+	if err != nil {
+		return err
+	}
+	logrus.Debugf("Image %s was created.", img.Name)
+
+	useImageImport, err := isImageImportSupported(cloud)
+	if err != nil {
+		return err
+	}
+
+	if useImageImport {
+		logrus.Debugf("Using Image Import API to upload RHCOS to the image %v with ID %v", img.Name, img.ID)
+		stageRes := imagedata.Stage(conn, img.ID, f)
+		if stageRes.Err != nil {
+			return err
+		}
+		logrus.Debugf("The data was uploaded.")
+
+		logrus.Debugf("Begin image import for the image %v with ID %v", img.Name, img.ID)
+		co := imageimport.CreateOpts{
+			Name: imageimport.GlanceDirectMethod,
+		}
+		importRes := imageimport.Create(conn, img.ID, co)
+		if importRes.Err != nil {
+			return err
+		}
+		logrus.Debugf("Image import started.")
+
+		// Image import is an asynchronous operation, so we have to wait until the image becomes "active"
+		const numRetries = 5000
+		const timeSleepSeconds = 15
+		for i := 0; i < numRetries; i++ {
+			getRes, err := images.Get(conn, img.ID).Extract()
+			if err != nil {
+				return err
+			}
+			if getRes.Status == images.ImageStatusActive {
+				break
+			}
+			time.Sleep(timeSleepSeconds * time.Second)
+		}
+
+		logrus.Debugf("Image import finished.")
+	} else {
+		// Use classic legacy upload that doesn't support image conversion
+		logrus.Debugf("Using legacy API to upload RHCOS to the image %v with ID %v", img.Name, img.ID)
+		res := imagedata.Upload(conn, img.ID, f)
+		if res.Err != nil {
+			return err
+		}
+		logrus.Debugf("The data was uploaded.")
+	}
+
+	return nil
+}
+
+// isImageImportSupported checks if we can use Image Import mechanism for image uploading
+func isImageImportSupported(cloud string) (bool, error) {
+	// More information about the Discovery API:
+	// https://docs.openstack.org/api-ref/image/v2/?expanded=#image-service-info-discovery
+	logrus.Debugln("Checking if the image import mechanism is supported")
+
+	opts := clientconfig.ClientOpts{
+		Cloud: cloud,
+	}
+
+	conn, err := clientconfig.NewServiceClient("image", &opts)
+	if err != nil {
+		return false, err
+	}
+
+	s, err := imageimport.Get(conn).Extract()
+	if err != nil {
+		// ErrDefault404 means that image discovery API is not available for the cloud
+		if _, ok := err.(gophercloud.ErrDefault404); ok {
+			return false, nil
+		}
+		return false, err
+	}
+
+	// Next two checks are just to make sure the response data was not corrupted
+	if s.ImportMethods.Description != "Import methods available." {
```

Comment:
I like the check just below this; so much so, that this check on "description" looks redundant. Moreover, relying on a strict check of a descriptive field looks brittle to me.

I'd remove the check on "Description", unless there's some additional reason for keeping it

Reply:
Agree that there is no need to have this check. I added it because this is a constant string and it's a part of the API https://docs.openstack.org/api-ref/image/v2/?expanded=import-methods-and-values-discovery-detail#image-service-info-discovery